File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/io/TransactionalOutputStream.java
package app.zxtune.io;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;

import app.zxtune.Log;

public class TransactionalOutputStream extends OutputStream {
  private static final String TAG = TransactionalOutputStream.class.getName();

  private final File target;
  private final File temporary;
  private OutputStream delegate;
  private boolean confirmed = false;

  public TransactionalOutputStream(File target) throws IOException {
    Log.d(TAG, "Write cached file %s", target.getAbsolutePath());
    this.target = target;
    this.temporary = new File(target.getPath() + "~" + Integer.toString(hashCode()));
    if (target.getParentFile().mkdirs()) {
      Log.d(TAG, "Created cache dir");
    }
    delegate = new FileOutputStream(temporary);
  }

  @Override
  public void close() throws IOException {
    delegate.close();
    delegate = null;
    if (confirmed) {
      if (!Io.rename(temporary, target)) {
        final String msg = "Failed to rename " + temporary + " to " + target;
        throw temporary.delete()
          ? new IOException(msg)
          : new IOException(msg, new IOException("Failed to delete temporary " + temporary));
      }
    } else if (!temporary.delete()) {
      throw new IOException("Failed to cleanup unconfirmed " + temporary);
    }
  }

  @Override
  public void flush() throws IOException {
    try {
      delegate.flush();
      confirmed = true;
    } catch (IOException e) {
      confirmed = false;
      throw e;
    }
  }

  @Override
  public void write(byte[] b) throws IOException {
    try {
      delegate.write(b);
    } catch (IOException e) {
      confirmed = false;
      throw e;
    }
  }

  @Override
  public void write(byte[] b, int off, int len) throws IOException {
    try {
      delegate.write(b, off, len);
    } catch (IOException e) {
      confirmed = false;
      throw e;
    }
  }

  @Override
  public void write(int b) throws IOException {
    try {
      delegate.write(b);
    } catch (IOException e) {
      confirmed = false;
      throw e;
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/io/TransactionalOutputStream.kt
package app.zxtune.io

import app.zxtune.Logger
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream

private val LOG = Logger(TransactionalOutputStream::class.java.name)

class TransactionalOutputStream(private val target: File) : OutputStream() {
    private val temporary: File
    private var delegate: OutputStream?
    private var confirmed = false

    init {
        LOG.d { "Write cached file ${target.absolutePath}" }
        temporary = File("${target.path}~${hashCode()}")
        if (target.parentFile!!.mkdirs()) {
            LOG.d { "Created cache dir" }
        }
        delegate = FileOutputStream(temporary)
    }

    override fun close() {
        delegate!!.close()
        delegate = null
        if (confirmed) {
            commit()
        } else if (!temporary.delete()) {
            throw IOException("Failed to cleanup unconfirmed $temporary")
        }
    }

    private fun commit() {
        if (!Io.rename(temporary, target)) {
            val msg = "Failed to rename $temporary to $target"
            throw if (temporary.delete()) {
                IOException(msg)
            } else {
                IOException(msg, IOException("Failed to delete temporary $temporary"))
            }
        }
    }

    override fun flush() {
        onDelegate { flush() }
        confirmed = true
    }

    override fun write(b: ByteArray) = onDelegate {
        write(b)
    }

    override fun write(b: ByteArray, off: Int, len: Int) = onDelegate {
        write(b, off, len)
    }

    override fun write(b: Int) = onDelegate {
        write(b)
    }

    private fun onDelegate(op: OutputStream.() -> Unit) =
        try {
            op(delegate!!)
        } catch (e: IOException) {
            confirmed = false
            throw e
        }
}